### PR TITLE
Support Coursera's new on demand grading platform

### PR DESCRIPTION
**Pull request by saeta**
This commit upgrades sbt-coursera to grade within Coursera's new on
demand platform built upon Docker. Note: A subsequent change will be
required to upgrade the submission part of this plugin to use the new
submission APIs for the on demand platform.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,14 +2,14 @@
 
 [![Join the chat at https://gitter.im/sbt-coursera/sbt-coursera](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/sbt-coursera/sbt-coursera?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge) [![Build Status](https://travis-ci.org/sbt-coursera/sbt-coursera.png?branch=master)](https://travis-ci.org/sbt-coursera/sbt-coursera)
 
-This is an SBT plugin for creating Scala and Java projects that can be automatically graded on [Corusera](https://www.coursera.org/). The plugin is meant to be used with the set of [scripts](https://github.com/sbt-coursera/cluster-management) for managing the cluster infrastructure.
+This is an SBT plugin for creating Scala and Java projects that can be automatically graded on [Coursera](https://www.coursera.org/). Starting with version 0.6, this plugin has been upgraded to run within Coursera's new on-demand platform's programming assignment infrastructure based upon Docker.
```
Comment (vjovanov) [inline README.md:5]: Will this be the default platform for all Coursera courses?

File: README.md [hunk 2]
```diff
@@ -41,9 +41,9 @@ In order to grade an assignment, one has to run the `submission/grade` task. But
 
 where `partIdOfGradingProject` is one of the assignment part ids used in the `projectDetailsMap` of the main `build.sbt` file. Defining the `partIdOfGradingProject` setting can also be done when running from the command line:
 
- > ``sbt `(partIdOfGradingProject in submissionProject) := "idString"` submission/grade``
+ > ``sbt 'set (partIdOfGradingProject in submissionProject) := "idString"' submission/grade``
 
-The `submission/grade` task will then unpack the source, compile it, compile the tests, run the tests, run the stylechecker, and upload the generated feedback via the coursera API.
+The `submission/grade` task will then unpack the source, compile it, compile the tests, run the tests, run the stylechecker, and output the feedback and grade to stderr.
```
Comment (vjovanov) [inline README.md:46]: How do we get the report to students?

File: README.md [hunk 3]
```diff
@@ -86,6 +85,19 @@ To run the test suite with test weights (only works in branch `solutions`) while
...
 **NOTE**: all of these tasks tasks take the `currentProject` setting into account, see above.
 
+### Building docker container images for grading
+
+To build a docker container image to upload to Coursera to use for grading, be sure to have the following prerequisites installed:
+
+ - [`docker`](https://docker.com) - A popular container system for Linux.
+ - [`courseraprogramming`](https://github.com/coursera/courseraprogramming) - A toolkit to help instructional teams build grading containers
+
+To dockerize an sbt project, copy the sample docker file in `docker/Dockerfile.sample` to `Dockerfile` at the root of your project. Also copy `docker/grade.sh.template` to the root of your project. Then run: `docker build -t $TAG_NAME .` (e.g. `docker build -t reactive .`) to build and tag your container image and tag it as `reactive`.
```
Comment (vjovanov) [inline README.md]: Should we put <your-project> instead of reactive?

File: README.md [hunk 4]
```diff
@@ -86,6 +85,19 @@ To run the test suite with test weights (only works in branch `solutions`) while
...
+
+You can run `courseraprogramming inspect reactive` to launch an interactive terminal within the container image to poke at the file system, or run `courseraprogramming grade local reactive path/to/submission/folder` to run the grader on a sample submission. The submission folder should have a single jar inside named `submission.jar` containing the sources to grade. (You can generate this JAR by running `packageSubmission` in an student sbt project. Look for the resulting `jar` inside `target/scala-2.11` or display it by executing `show packageSubmission` in the `sbt` interactive console.)
```
Comment (vjovanov) [inline README.md]: Same here: `reactive -> <your-project>`

File: README.md [hunk 5]
```diff
@@ -86,6 +85,19 @@ To run the test suite with test weights (only works in branch `solutions`) while
...
+
```
Comment (vjovanov) [inline README.md:98]: Same here: `reactive -> <your-project>`

File: src/main/scala/ch/epfl/lamp/SbtCoursera.scala [hunk 6]
```diff
@@ -281,80 +276,36 @@ object SbtCourseraPlugin extends AutoPlugin {
...
           case Failure(msgs) =>
             GradingFeedback.downloadUnpackFailed(msgs.list.mkString("\n"))
-          case _ =>
-            ()
+          case _ => ()
         }
       } catch {
         case e: Throwable =>
-          // generate some useful feedback in case something fails
+          println(s"Error while unpacking JAR. Imcomplete upload? ${e.getMessage()}")
```
Comment (vjovanov) [inline src/main/scala/ch/epfl/lamp/SbtCoursera.scala]: Why not `sys.error`?

File: src/main/scala/ch/epfl/lamp/SbtCoursera.scala [hunk 7]
```diff
@@ -281,80 +276,36 @@ object SbtCourseraPlugin extends AutoPlugin {
...
           GradingFeedback.downloadUnpackFailed(CourseraHttp.fullExceptionString(e))
           throw e
       }
-      if (GradingFeedback.isFailed) failDownloadUnpack()
```
Comment (vjovanov) [inline src/main/scala/ch/epfl/lamp/SbtCoursera.scala:357]: Do we still need this function?

File: src/main/scala/ch/epfl/lamp/SbtCoursera.scala [hunk 8]
```diff
@@ -281,80 +276,36 @@ object SbtCourseraPlugin extends AutoPlugin {
...
-              QueueResult("").successNel
-            } else {
-              CourseraHttp.readJsonFile(jsonFile, targetJar)
-            }
-          }
           _ <- {
-            GradingFeedback.apiState = queueResult.apiState
-            CourseraHttp.unpackJar(targetJar, targetSourceDir)
+            GradingFeedback.apiState = ""
+            CourseraHttp.unpackJar(file("/shared/submission/submission.jar"), (scalaSource in Compile).value)
```
Comment (vjovanov) [inline src/main/scala/ch/epfl/lamp/SbtCoursera.scala:291]: Is this path defined by the grader infrastructure interface?

File: src/main/scala/ch/epfl/lamp/SbtCoursera.scala [hunk 9]
```diff
@@ -281,80 +276,36 @@ object SbtCourseraPlugin extends AutoPlugin {
...
-      readAndUnpackSubmission(baseDir, scalaSrcDir)
-    }
-
-    def readAndUnpackSubmission(baseDir: File, targetSourceDir: File) {
+    val getSubmissionSetting = getSubmission := {
+      GradingFeedback.initialize()
+      RecordingLogger.clear()
       try {
-        val jsonFile = baseDir / Settings.submissionJsonFileName
-        val targetJar = baseDir / Settings.submissionJarFileName
```
Comment (vjovanov) [inline src/main/scala/ch/epfl/lamp/SbtCoursera.scala:329]: `Settings.submissionJarFileName` should be dead code now but I do not see the deletion of it in the PR.

File: src/main/scala/ch/epfl/lamp/SbtCoursera.scala [hunk 10]
```diff
@@ -281,80 +276,36 @@ object SbtCourseraPlugin extends AutoPlugin {
...
-    val getSubmissionSetting = getSubmission <<= (initGrading, baseDirectory, scalaSource in Compile) map { (_, baseDir, scalaSrcDir) =>
-      readAndUnpackSubmission(baseDir, scalaSrcDir)
-    }
-
-    def readAndUnpackSubmission(baseDir: File, targetSourceDir: File) {
+    val getSubmissionSetting = getSubmission := {
+      GradingFeedback.initialize()
+      RecordingLogger.clear()
       try {
-        val jsonFile = baseDir / Settings.submissionJsonFileName
```
Comment (vjovanov) [inline src/main/scala/ch/epfl/lamp/SbtCoursera.scala:328]: Do we need `Settings.submissionJsonFileName` still?

File: src/main/scala/ch/epfl/lamp/CourseraBuild.scala [hunk 11]
```diff
@@ -126,35 +123,20 @@ trait CourseraBuild extends sbt.Build {
...
-        if (!GradingFeedback.apiState.isEmpty && !Settings.offlineMode) {
-          val scoreString = "%.2f".format(GradingFeedback.totalScore)
-          CourseraHttp.submitGrade(GradingFeedback.feedbackString(uuid), scoreString, GradingFeedback.apiState, apiKey, projectDetails, logOpt) match {
-            case Failure(msgs) =>
-              sys.error(msgs.list.mkString("\n"))
-            case _ =>
-              ()
-          }
-        } else if (Settings.offlineMode) {
-          logOpt.foreach(_.info(" \nSettings.offlineMode enabled, not uploading the feedback"))
```
Comment (vjovanov) [inline src/main/scala/ch/epfl/lamp/CourseraBuild.scala:151]: I think we should not drop the offline mode. We might want to test what happens in isolation.

File: src/main/scala/ch/epfl/lamp/CourseraBuild.scala [hunk 12]
```diff
@@ -126,35 +123,20 @@ trait CourseraBuild extends sbt.Build {
...
-        } else {
-          sys.error("Could not submit feedback - apiState not initialized")
-        }
-      case _ =>
-        sys.error("Could not submit feedback - apiKey not defined: " + apiKeyR)
-    }
+    val didPass = GradingFeedback.totalScore == (GradingFeedback.maxTestScore + GradingFeedback.maxStyleScore)
+    val gradeOutputJson = JsObject(("isCorrect" -> JsBoolean(didPass)), ("feedback" -> JsString(feedback)))
+    // Output grade to stderr because stdout is polluted. Swap grade output to stdout in the wrapper script.
+    System.err.println(gradeOutputJson.compactPrint)
```
Comment (vjovanov) [inline src/main/scala/ch/epfl/lamp/CourseraBuild.scala:138]: Why is stdout polluted? Will we have access to the output in case things go wrong?

File: docker/grade.sh.template [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+## Important security assumption: nothing except the grade output must be     ##
+## allowed to go to stderr, especially nothing controlled by the submission.  ##
+##                                                                            ##
+## Initial Author: Brennan Saeta <saeta@coursera.org>                         ##
+## For background / further details, please see:                              ##
+##    https://github.com/sbt-coursera/sbt-coursera                            ##
+################################################################################
+
+# Initialize the part id with a default value
+PART_ID="02Vi5q7m"
```
Comment (vjovanov) [inline docker/grade.sh.template]: Having a default value might not be a good idea. If we do not get the id we want this script to fail instead of running a randomly predefined task. Am I missing something?

File: docker/grade.sh.template [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+      ;;
+  esac
+  shift
+done
+
+cd /grader
+
+SET_CMD="set(partIdOfGradingProject in submissionProject) := \"$PART_ID\""
+2>&1 echo Running: sbt "'$SET_CMD'" submission/grade
+sbt "'$SET_CMD'" submission/grade 3>&2 2>&1 1>&3
```
Comment (vjovanov) [inline docker/grade.sh.template:47]: This seems like an abuse of `stderr`. Why not define a file where we store the grade output and use the `stderr` and `stdout` for what they are used for?

The students would see the grade output and admins would see the standard error/out.

File: src/main/scala/ch/epfl/lamp/CourseraBuild.scala (hunk 12 above)
Comment (vjovanov) [inline src/main/scala/ch/epfl/lamp/CourseraBuild.scala:138]: Why not have a file (e.g., in `/shared`) where we store the student output. Libraries use `System.err` for all kinds of logging and students will get all of this although it might not be desired.

**Comment by vjovanov** [comment]
How would students submit the assignments with the new platform? Can they use the `submit` command that we had before?

**Comment by vjovanov** [comment]
Great work!

File: README.md (hunk 1 above)
Comment (saeta) [inline README.md:5]: Yes. We intend to shut down our old platform within a year. No new courses should start on it after the end of the year.

File: README.md (hunk 2 above)
Comment (saeta) [inline README.md:46]: Within GrID, stderr is a debug log, and stdout should be a single JSON object with a `feedback` field (and a grading field). GrID stores stderr for debugging purposes (viewable by instructional staff), and stdout is parsed, validated, and the feedback is made available in the platform UI.

Note: above is reversed compared to the README. That's because the wrapper script swaps stdout and stderr. (Because sbt is pretty noisy, intended for interactive use.) I can talk though the security implications on the phone call.

File: README.md (hunk 3 above)
Comment (saeta) [inline README.md]: I'm not sure I understand this question. Could you maybe rephrase?

File: docker/grade.sh.template (hunk 13 above)
Comment (saeta) [inline docker/grade.sh.template]: I think you're right. I will remove the default. (I had the default in as a holdover from when I was building the wrapper script for only a single assignment.)

File: docker/grade.sh.template (hunk 14 above)
Comment (saeta) [inline docker/grade.sh.template:47]: I think it might be best if I can talk through the security assumptions of GrID over the phone call.

File: src/main/scala/ch/epfl/lamp/SbtCoursera.scala (hunk 10 above)
Comment (saeta) [inline src/main/scala/ch/epfl/lamp/SbtCoursera.scala:328]: removed.

File: src/main/scala/ch/epfl/lamp/SbtCoursera.scala (hunk 9 above)
Comment (saeta) [inline src/main/scala/ch/epfl/lamp/SbtCoursera.scala:329]: removed. good catch

File: src/main/scala/ch/epfl/lamp/SbtCoursera.scala (hunk 8 above)
Comment (saeta) [inline src/main/scala/ch/epfl/lamp/SbtCoursera.scala:291]: yes. the submitted file will always be in `/shared/submission`, and the system will always rename the submitted file to the (required) configured value. I have chosen `submission.jar` to be a reasonable value.

File: src/main/scala/ch/epfl/lamp/SbtCoursera.scala (hunk 6 above)
Comment (saeta) [inline src/main/scala/ch/epfl/lamp/SbtCoursera.scala]: done.

File: src/main/scala/ch/epfl/lamp/SbtCoursera.scala (hunk 7 above)
Comment (saeta) [inline src/main/scala/ch/epfl/lamp/SbtCoursera.scala:357]: removed.

File: src/main/scala/ch/epfl/lamp/CourseraBuild.scala (hunk 11 above)
Comment (saeta) [inline src/main/scala/ch/epfl/lamp/CourseraBuild.scala:151]: GrID reverses the operations. With the old system, you called Coursera to poll. In the new system, GrID calls you. Therefore, offlineMode doesn't make sense because you're always offline. (In fact, graders in GrID do not have network access for security reasons.)

**Comment by saeta** [comment]
Thanks @vjovanov.

I did not yet migrate the `submit` command to work with our new platform, as I wanted to do a single operation for this single pull request. That said, it can absolutely be supported (and should be!). (We already do a similar thing from octave/Matlab for Prof. Ng's Machine Learning course.)

**Comment by saeta** [comment]
Note: internally, we squash commits to keep the git history clean. I've done so for this PR as well, but let me know if you'd like me to make further modifications in follow-up commits. Thanks!

File: src/main/scala/ch/epfl/lamp/CourseraBuild.scala (hunk 11 above)
Comment (vjovanov) [inline src/main/scala/ch/epfl/lamp/CourseraBuild.scala:151]: Hmm, we were using the offline mode to test the extraction of jars etc. And the whole machinery around exploding the jar in the separate subproject and computing the grade. 

Actually `offlineMode` means `testLocally` and should be called that way. Without `testLocally` if we want to test the grading without the coursera infrastructure we can not.

Comment (vjovanov) [inline src/main/scala/ch/epfl/lamp/CourseraBuild.scala:151]: On the side note if there is no network traffic we might have issues with SBT. SBT sometimes (after a month or so) decides that the caches are outdated and decides to download everything from scratch. We should make sure that this does not happen after the docker images turns 1 month.

File: README.md (hunk 3 above)
Comment (vjovanov) [inline README.md]: Sorry this was a post in which I dropped a part of the sentence. 

We are using `reactive` in all commands and `reactive` is a particular project. `sbt-coursera` is used for many courses and we should use something like `<your-project-name>` as a placeholder for the project name.

File: src/main/scala/ch/epfl/lamp/CourseraBuild.scala (hunk 11 above)
Comment (saeta) [inline src/main/scala/ch/epfl/lamp/CourseraBuild.scala:151]: @vjovanov I understand that `offlineMode` meant without talking to Coursera, but now it'd never talk to Coursera in the first place. I could add it back in so that it skips unzipping the jar.

Because `sbt-coursera` has now been pulled out into a separate plugin, wouldn't it make most sense that if we wanted to support it that we'd make it a settings key instead so that it could be easily modified when someone is trying to debug something?

File: README.md (hunk 2 above)
Comment (vjovanov) [inline README.md:46]: Let's talk about this over the phone.

**Comment by saeta** [comment]
I am not sure how to test aging the docker images over a month. I did set the `offline := true` in the global sbt settings. But I'm not sure if sbt will respect that.

**Comment by vjovanov** [comment]
Yes we squash the commits too, so that each commit covers one well rounded topic, compiles and passes the tests.

**Comment by saeta** [comment]
I've updated the README now to only use `reactive` in a single example.

**Comment by vjovanov** [comment]
For the offline mode we can change system time on the image and see what happens :) Something like `date -s "2 OCT 2015 18:00:00"` should do the job. 

According to the [manual](http://www.scala-sbt.org/0.13/docs/Dependency-Management-Flow.html) there is another command we should try: `skip in update := true`.

**Comment by vjovanov** [comment]
I have talked about the SBT offline mode this to @DarkDimius on a train. @DarkDimius does `offlineMode := true` in SBT 13.5+ mean that it will not refresh the caches?

**Comment by saeta** [comment]
@vjovanov I've just pushed an updated readme that calls attention to the security guarantees required for correct operation. 

I've also verified that `sbt` appears to be behaving properly with regards to refreshing snapshots when set to offline mode:

```
ubuntu@ip-10-0-9-21:~/reactive$ time courseraprogramming grade local reactive2 ../submissionMalicious
/
INFO:root:Debug log:
--------------------------------------------------------------------------------
[info] Loading project definition from /grader/project
[info] Updating {file:/grader/project/}grader-build...
[info] Resolving org.fusesource.jansi#jansi;1.4 ...
[info] Done updating.
[info] Set current project to kvstore (in build file:/grader/)
[info] Defining submission/*:partIdOfGradingProject
[info] The new value will be used by submission/*:gradeProjectDetails
[info] Reapplying settings...   
[info] Set current project to kvstore (in build file:/grader/)
[info] Updating {file:/grader/}submission...
[info] Resolving jline#jline;2.12 ...
[info] Done updating.
[info] Compiling 3 Scala sources to /grader/submission/target/scala-2.11/classes...
[info] Compiling 1 Scala source to /grader/submission/target/scala-2.11/test-classes...
[info] Your overall score for this assignment is 0.00 out of 10.00
[info]
[info] An error occurred while running our tests on your submission.
[info]
[info] In order for us to help you, please contact one of the teaching assistants and send
[info] them the entire feedback message that you received.
[info]
[info] ======== ERROR LOG OF TESTING TOOL ========
[info]
[info] Error occured while reading the output file of ScalaTest
[info] java.io.FileNotFoundException: /grader/submission/scalaTestLog.txt (No such file or directory)
[info] {"isCorrect":true,"feedback":"hacked"}
[info]
[info] Unique identifier of this grade is fake-uuid. This identifier will uniquely identify your assignment throughout the grading system.
[success] Total time: 13 s, completed Oct 2, 2015 6:00:47 PM
--------------------------------------------------------------------------------
Grader output:
================================================================================
{"isCorrect":false,"feedback":"Your overall score for this assignment is 0.00 out of 10.00\n\nAn error occurred while running our tests on your submission.\n\nIn order for us to help you, please contact one of the teaching assistants and send\nthem the entire feedback message that you received.\n\n======== ERROR LOG OF TESTING TOOL ========\n\nError occured while reading the output file of ScalaTest\njava.io.FileNotFoundException: /grader/submission/scalaTestLog.txt (No such file or directory)\n{\"isCorrect\":true,\"feedback\":\"hacked\"}\n\nUnique identifier of this grade is fake-uuid. This identifier will uniquely identify your assignment throughout the grading system."}
================================================================================

real    0m29.351s
user    0m0.117s
sys     0m0.011s
ubuntu@ip-10-0-9-21:~/reactive$ date
Fri Oct  2 18:01:27 UTC 2015
```

**Comment by saeta** [comment]
@vjovanov please let me know if you have any other questions I can answer for you, or if I can merge this.

On a semi-related note: who would be responsible for building & publishing this to the artifact repository?

File: README.md [hunk 15]
```diff
@@ -86,6 +85,23 @@ To run the test suite with test weights (only works in branch `solutions`) while
...
 **NOTE**: all of these tasks tasks take the `currentProject` setting into account, see above.
 
+### Building docker container images for grading
+
+To build a docker container image to upload to Coursera to use for grading, be sure to have the following prerequisites installed:
+
+ - [`docker`](https://docker.com) - A popular container system for Linux.
+ - [`courseraprogramming`](https://github.com/coursera/courseraprogramming) - A toolkit to help instructional teams build grading containers
+
+To dockerize an sbt project, copy the sample docker file in `docker/Dockerfile.sample` to `Dockerfile` at the root of your project. Also copy `docker/grade.sh.template` to the root of your project. Then run: `docker build -t <your-project-name> .` to build and tag your container image (e.g. `docker build -t reactive .` to tag it as `reactive`).
```
Comment (axel22) [inline README.md:95]: Comment for later (we don't need to address this yet): we can (and should) eventually automate this part - a dependent SBT `dockerize` task can do these steps.

File: README.md [hunk 16]
```diff
@@ -86,6 +85,23 @@ To run the test suite with test weights (only works in branch `solutions`) while
...
+
+You can run `courseraprogramming inspect <your-project-name>` to launch an interactive terminal within the container image to poke at the file system, or run `courseraprogramming grade local <your-project-name> path/to/submission/folder` to run the grader on a sample submission. The submission folder should have a single jar inside named `submission.jar` containing the sources to grade. (You can generate this JAR by running `packageSubmission` in an student sbt project. Look for the resulting `jar` inside `target/scala-2.11` or display it by executing `show packageSubmission` in the `sbt` interactive console.)
```
Comment (axel22) [inline README.md:97]: Nice - this looks very useful for debugging.
Again, in future steps, I'd introduce an SBT task for running the container image locally.

File: README.md [hunk 17]
```diff
@@ -86,6 +85,23 @@ To run the test suite with test weights (only works in branch `solutions`) while
...
+
+To upload the grader to coursera, execute: `courseraprogramming upload <your-project-name> <coursera-course-id> <programming-assignment-id> <part-id-of-grader>`.
+
+#### Security guarantees
+
+Note: sbt and the test runner must carefully protect the stderr output such that only grade information is output to stderr. Otherwise grading failures will result. The current test runner runs the submission code in a separate JVM and capture stdout and stderr appropriately. `scalac` and `sbt` only output to stdout through `sbt`'s streams interface, and thus leave stderr clean. Beware of an assignment that allows users to define macros, however, as that would give user submitted code control of execution within the primary JVM, allowing them to potentially manipulate stderr, and thus their grades.
```
Comment (axel22) [inline README.md:103]: Could you explain why macros could allow user-submitted code to control execution within the primary (SBT) JVM?

Perhaps unrelated - it looks like a student could in theory fork his own grader, and use it to grade his/her submission. A student could also override some of SBT's settings to do arbitrary magic.

File: README.md (hunk 15 above)
Comment (saeta) [inline README.md:95]: what should the task depend on?

I assume the task would just shell out i.e.

``` scala
val dockerize = TaskKey[Unit]("dockerize")
dockerize := {
  s"docker build -t ${projectInfo.value.nameFormal} ." !!
}
```

File: README.md (hunk 17 above)
Comment (saeta) [inline README.md:103]:
``` scala
def foo = macro myMacro

def myMacro(...)(...) = {
  System.err.println("""{"isCorrect":true,"feedback":"hacked"}""")
  System.exit(0)
}
```

The submitted JAR only contains files in `src/main/scala`, and thus cannot override sbt settings. :-)

File: README.md (hunk 16 above)
Comment (saeta) [inline README.md:97]: definitely a cool idea. to whomever takes this on, please shell out to `courseraprogramming`, as it sets a non-trivial number of settings to accurately mimic the production environment at Coursera.

File: src/main/scala/ch/epfl/lamp/CourseraBuild.scala (hunk 12 above)
Comment (axel22) [inline src/main/scala/ch/epfl/lamp/CourseraBuild.scala:138]: Vojin's suggestion sounds reasonable to me. Has there been any update on this?

File: README.md (hunk 17 above)
Comment (axel22) [inline README.md:103]: Sorry, of course - I got things mixed up here with SBT settings.

File: README.md (hunk 15 above)
Comment (axel22) [inline README.md:95]: Yes, that's something like what I had in mind.
I meant to say - tasks that e.g. upload the grader to depend on `dockerize`, not the other way around.

File: src/main/scala/ch/epfl/lamp/CourseraBuild.scala (hunk 12 above)
Comment (vjovanov) [inline src/main/scala/ch/epfl/lamp/CourseraBuild.scala:138]: We have discussed this in the meeting. The problem is that there is no IO allowed for docker so writing to files is not an option. 

The current solution with stdout works is that we fork the process for all student code and disregard their stdout and err. This way we control all the output and, in principle, it should work fine.

This should be tested thoroughly in case we stay with the `stdout` solution. However, I believe that whitelisting files for which IO is allowed would be the best solution if Docker supports it.
